scripts/ci/remote-build-artifacts.sh

```diff
@@ -0,0 +1,184 @@
+
+    remoteChecksumFilePath="${s3BuildDir}/${frameworkName}/${checksumFileName}"
+    copiedRemoteChecksumFile="./${checksumFileName}"
+
+    if isFileExistsOnS3 "${remoteChecksumFilePath}"; then
+        aws s3 cp "${remoteChecksumFilePath}" "${copiedRemoteChecksumFile}"
+
+        if isFileExistsAtPath "${copiedRemoteChecksumFile}"; then
+            remoteChecksum=$(cat ${copiedRemoteChecksumFile})
+            
+            echo
+            echo "Checksum for a remote build product: ${remoteChecksum}"
+            echo "Checksum for a local build product:  ${localChecksum}"
+
+            if [[ "${localChecksum}" == "${remoteChecksum}" ]]; then
+                echo "${frameworkName} at ${s3BuildDir} was not changed, update is not needed"
+                performCleanup
+                return
+            else
+                echo "There's a new version of ${framework}, update is needed"
+            fi
+        fi
+    else
+        echoerr "Checksum file does not exist at ${remoteChecksumFilePath}"
+        exit "$ERROR_FILE_DOES_NOT_EXIST"
+    fi
+
+    # We implement "mutual exclusion" with files we can treat as flags.
+    # Each actor (repository) which wants to work with AWS S3, sets up a flag (creates an empty marker file).
+    # Other actors (repositories) can easily see the flags.
+    # When `mapbox-vision-ios` repo wants to pull the new verison of framework, it does the following:
+    # 1. raise the flag (creates a pullMarkerFlag file)
+    # 2. when there's no syncMarkerFlag, we start pulling process
+    # 3. when the pulling is completed, we lower the flag (remove pullMarkerFlag file)
+    #
+    # What we do in case when there's a syncMarkerFlag (step 2 above)?
+    # - we lower our flag (removes pullMarkerFlag file)
+    # - we wait until sync flag is lowered (syncMarkerFlag is removed)
+    # - then we raise our flag (create pullMarkerFlag file)
+    #
+    # The same behaviour is applicable for other actors (including mapbox-vision repo).
+
+    echo
+    echo "Checking if build artifacts are currently being updated"
+    echo
+    syncMarkerFilePath="${s3BuildDir}/${frameworkName}/${syncMarkerFileName}"
+    pullMarkerFilePath="${s3BuildDir}/${frameworkName}/${pullMarkerFileName}"
+
+    while : ; do
+        echo "Putting a pull marker to indicate we want to pull latest build artifacts"
+        createEmptyFileOnS3 "${pullMarkerFilePath}"
+
+        if isFileExistsOnS3 "${syncMarkerFilePath}"; then
+            echo "Sync marker was found. That means uploading of build artifacts is in progress"
+            echo "Removing the pull marker"
+            aws s3 rm "${pullMarkerFilePath}" --quiet
+        else
+            echo "Sync marker was not found. That means we can start pulling process"
+            break
+        fi
+
+        echo "Retrying in several seconds"
+        echo
+        sleep $[ ( $RANDOM % 20 )  + 1 ]s # 1-20 seconds random sleep, time interval to check flag next time
```

Comment:
maybe we should add a max number of tries or summary duration of the tries? there can be problems with network connection, or with the mapbox-vision repository, or something else